UPDATE: Code scanning with GitHub CodeQL

a-a-ron opened the pull request:
The purpose of this PR is to see how we can increase the completion rate and improve the user experience. Below is the August data for this module:

| Module | Completion Rate | # completed / # started |
|--------|--------|--------|
| [Code scanning with GitHub CodeQL](https://learn.microsoft.com/en-us/training/modules/code-scanning-with-github-codeql/?ns-enrollment-type=Collection&ns-enrollment-id=rqymc6yw8q5rey)| 15.87% | 33/208 |

**For reference:** See the [parent issue](https://github.com/githubpartners/microsoft-learn/issues/35) for context

## Module file structure
![Screen Shot 2022-09-28 at 1 54 00 PM](https://user-images.githubusercontent.com/6351798/192876845-fd2b995f-ca4a-4019-991c-ca1641745dac.png)

## Items to review for this module

We need to conduct a performance review on each of these modules to see how we can increase the completion rates, improve the user experience, and maintain the level of quality that we set for our learners. 

### Items for consideration
- **Thoroughly review the module content for flow, narrative, and length**
  - Review use of images, tables, and other visual elements
  - Does the content read like a doc or like a conversation guide?
  - Does the estimated time to complete each unit remain accurate?
- **Review the introduction unit** 
  - Does it capture your attention?
  - Does it contain repetitive content from the course title page?
  - Do the learning objectives/scope align with the content? 
  - Does the target audience match the content?
  - Do we accurately describe the module and learning objectives? (Is it possible that learners start the course thinking it's one thing to find out it's another?) 
- **Review the module exercise**
  - Does it work?
  - Does it enforce a learning objective?
  - How is it placed in the module? Should we move it around?
  - Do we need other exercises?
- **Review the questions at the end of the module**
  - Do the questions test on content taught in the module?
  - Do the questions follow best practices for item writing?

## Estimated Work Effort
| Task | Hours | 
|--------|--------|
| Detailed module review | 18 hours | 
| Action items as identified from review | TBD |

a-a-ron committed:
Update 1-introduction.yml

rmallorybpc commented:
## Plan of Action

### Launch page

- From the launch page, this is a long course.
- Could the module be broken into CodeQL and then Custom?
- Move to a new Custom module: Reference a custom CodeQL query, Learn how to use the CodeQL CLI to generate code scanning results and upload them to GitHub, and Implement custom build steps
   - Those launch bullet points seem to match the units: Customize your code scanning workflow with CodeQL - Part 1, Customize your code scanning workflow with CodeQL - Part 2, and Customize languages and builds for code scanning

### Introduction

- Interesting introduction. Same questions as launch page about the amount of content.
- Yes, the learning objectives/scope aligns with the content.

Content Units

- Could be better conversation. Unit 2 What is CodeQL?, does not continue the story from the intro.

rmallorybpc commented:
## Move Introduction information

This comment was used to save the moved content, copy and paste, and compare the content across the two files.

## Learning objectives

By the end of this module, you will be able to:

* Understand CodeQL and how it analyzes code
* Understand QL, a unique logic programming language
* Set up CodeQL based code scanning in a GitHub repository
* Reference a custom CodeQL query
* Configure the language matrix in a CodeQL workflow
* Learn how to use the CodeQL CLI to generate code scanning results and upload them to GitHub
* Implement custom build steps

## Prerequisites

- A GitHub enterprise account with a GitHub Advanced Security license
- Necessary permissions to administrate your repository
- Knowledge of GitHub Advanced Security's code scanning feature
- Knowledge of GitHub Actions

rmallorybpc committed:
Update 1-introduction.md

Moved the learning objectives and prerequisites to the launch page, and added a transition sentence.

rmallorybpc commented:
# New module order

## Introduction to Code scanning with GitHub CodeQL

1-introduction.md - edited
2-what-is-codeql.md
3-how-does-codeql-analyze-code.md
4-what-is-ql.md
Exercise - new exercise needed
12-summary.md - edited

# New module

## Code Scanning with GitHub CodeQL

5-code-scanning-codeql.md
9-use-codeql-cli.md
[Recommended hardware resources for running CodeQL](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/recommended-hardware-resources-for-running-codeql) - new unit
Exercise - new exercise needed
12-summary.md - edited

# New module

## Customize Code Scanning with GitHub CodeQL

1-introduction.md - edited
[Configuring the CodeQL workflow for compiled languages](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-the-codeql-workflow-for-compiled-languages) - new unit
6-customize-your-scanning-workflow-with-codeql.md
7-exercise-reference-codeql-query.md
8-customize-your-scanning-workflow-with-codeql-2.md
11-exercise-configure-language-matrix.md
10-custom-build-steps-for-code-scanning.md
[Troubleshooting the CodeQL workflow](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/troubleshooting-the-codeql-workflow) - new unit
Exercise(s) needed
12-summary.md

rmallorybpc committed:
Update 1-introduction.md

Added the learning objectives and prerequisites back

rmallorybpc commented:
# New module order

Proposed, the original module, [Code scanning with GitHub CodeQL](https://learn.microsoft.com/en-us/training/modules/code-scanning-with-github-codeql/), to be split into 3 modules.

## Introduction to Code scanning with GitHub CodeQL

1-introduction.md - edited existing unit

Code scanning using CodeQL provides an extensible method to automate vulnerability scanning across your organizations GitHub repositories.

Imagine that you are a senior developer at a start-up company specializing in health care software. Your flagship product is a Java-based web portal that allows physicians to manage patient records. A recent penetration test of this product revealed a number of serious vulnerabilities that could compromise patient information. The CIO has asked you to implement automated code vulnerability scanning. Because your code is already hosted in a private repository on GitHub, you have decided to use the code scanning feature with CodeQL. You will need to understand how the feature works to persuade other developers and management to use the feature. You will also need to understand the various configuration options and how to implement and maintain a code scanning pipeline to assist other developers at your company in configuring and deploying code scanning correctly.

In this module, you will learn about the CodeQL static analysis tool and how the code scanning feature in GitHub uses it to automate vulnerability scanning. You will also learn how to customize a code scanning workflow that uses CodeQL, how to include additional queries, and how to adapt your workflow to repositories that have multiple languages.

## Learning objectives

By the end of this module, you will be able to:

- Understand CodeQL and how it analyzes code
- Understand QL, a unique logic programming language
- Set up CodeQL based code scanning in a GitHub repository
- Reference a custom CodeQL query
- Configure the language matrix in a CodeQL workflow
- Learn how to use the CodeQL CLI to generate code scanning results and upload them to GitHub
- Implement custom build steps

## Prerequisites

- A GitHub enterprise account with a GitHub Advanced Security license
- Necessary permissions to administrate your repository
- Knowledge of GitHub Advanced Security's code scanning feature
- Knowledge of GitHub Actions

Next up, you'll learn how CodeQL is used by developers.

2-what-is-codeql.md - existing unit

CodeQL is the analysis engine used by developers to automate security checks, and by security researchers to perform variant analysis.

In CodeQL, code is treated like data. Security vulnerabilities, bugs, and other errors are modeled as queries that can be executed against databases extracted from code. You can run the standard CodeQL queries, written by GitHub researchers and community contributors, or write your own to use in custom analyses. Queries that find potential bugs highlight the result directly in the source file.

In this unit, you will learn about the CodeQL static analysis tool and how it uses databases, query suites and query language packs to perform variant analysis.

## Variant analysis

Variant analysis is the process of using a known security vulnerability as a seed to find similar problems in your code. It’s a technique that security engineers use to identify potential vulnerabilities, and ensure these threats are properly fixed across multiple codebases.

Querying code using CodeQL is the most efficient way to perform variant analysis. You can use the standard CodeQL queries to identify seed vulnerabilities, or find new vulnerabilities by writing your own custom CodeQL queries. Then, develop or iterate over the query to automatically find logical variants of the same bug that could be missed using traditional manual techniques.

## CodeQL databases

CodeQL databases contain queryable data extracted from a codebase, for a single language at a particular point in time. The database contains a full, hierarchical representation of the code, including a representation of the abstract syntax tree, the data flow graph, and the control flow graph.

Each language has its own unique database schema that defines the relations used to create a database. The schema provides an interface between the initial lexical analysis performed during the extraction process, and the actual complex analysis of the CodeQL query evaluator. The schema specifies, for instance, that there is a table for every language construct.

For each language, the CodeQL libraries define classes to provide a layer of abstraction over the database tables. This provides an object-oriented view of the data which makes it easier to write queries.

For example, in a CodeQL database for a Java program, two key tables are:

* The `expressions` table containing a row for every single expression in the source code that was analyzed during the build process.
* The `statements` table containing a row for every single statement in the source code that was analyzed during the build process.

The CodeQL library defines classes to provide a layer of abstraction over each of these tables (and the related auxiliary tables): `Expr` and `Stmt`.

## Query suites

CodeQL query suites provide a way of selecting queries, based on their filename, location on disk or in a QL pack, or metadata properties. Create query suites for the queries that you want to frequently use in your CodeQL analyses.

Query suites allow you to pass multiple queries to CodeQL without having to specify the path to each query file individually. Query suite definitions are stored in YAML files with the extension `.qls`. A suite definition is a sequence of instructions, where each instruction is a YAML mapping with (usually) a single key. The instructions are executed in the order they appear in the query suite definition. After all the instructions in the suite definition have been executed, the result is a set of selected queries.

### Default query suites

There are three default query suites for CodeQL:

- `code-scanning`: queries run by default in CodeQL code scanning on GitHub.
- `security-extended`: queries from `code-scanning`, plus extra security queries with slightly lower precision and severity.
- `security-and-quality`: queries from `code-scanning`, `security-extended`, plus extra maintainability and reliability queries.

## Query Language (QL) packs

QL packs are used to organize the files used in CodeQL analysis. They contain queries, library files, query suites, and important metadata.

The CodeQL repository contains QL packs for C/C++, C#, Java, JavaScript, Python, and Ruby. The CodeQL for Go repository contains a QL pack for Go analysis. You can also make custom QL packs to contain your own queries and libraries.

### QL pack structure

A QL pack must contain a file called `qlpack.yml` in its root directory. The other files and directories within the pack should be logically organized. For example:

* Queries are organized into directories for specific categories.
* Queries for specific products, libraries, and frameworks are organized into their own top-level directories.
* There is a top-level directory named `<owner>/<language>` for query library (`.qll`) files. Within this directory, `.qll` files should be organized into subdirectories for specific categories.

An example `qlpack.yml` file is shown below.

```yml
name: codeql/java-queries
version: 0.0.6-dev
groups: java
suites: codeql-suites
extractor: java
defaultSuiteFile: codeql-suites/java-code-scanning.qls
dependencies:
    codeql/java-all: "*"
    codeql/suite-helpers: "*"
```

3-how-does-codeql-analyze-code.md - existing unit

Implementing code scanning with CodeQL requires an understanding of how the tool analyzes code.

CodeQL analysis consists of three steps:

1. Preparing the code, by creating a CodeQL database.
2. Running CodeQL queries against the database.
3. Interpreting the query results.

In this unit, you will learn about the three phases of CodeQL analysis.

## Database creation

To create a database, CodeQL first extracts a single relational representation of each source file in the codebase.

For compiled languages, extraction works by monitoring the normal build process. Each time a compiler is invoked to process a source file, a copy of that file is made, and all relevant information about the source code is collected. This includes syntactic data about the abstract syntax tree and semantic data about name binding and type information.

For interpreted languages, the extractor runs directly on the source code, resolving dependencies to give an accurate representation of the codebase.

There is one extractor for each language supported by CodeQL to ensure that the extraction process is as accurate as possible. For multi-language codebases, databases are generated one language at a time.

After extraction, all the data required for analysis (relational data, copied source files, and a language-specific database schema, which specifies the mutual relations in the data) is imported into a single directory, known as a CodeQL database.

## Query execution

After you’ve created a CodeQL database, one or more queries are executed against it. CodeQL queries are written in a specially designed object-oriented query language called QL.

You can run the queries checked out from the CodeQL repo (or custom queries that you’ve written yourself) using the CodeQL for VS Code extension or the CodeQL CLI.

## Query results

The final step converts results produced during query execution into a form that is more meaningful in the context of the source code. That is, the results are interpreted in a way that highlights the potential issue that the queries are designed to find.

:::image type="content" source="../media/codeql-query-results.png" alt-text="Screenshot of CodeQL query results.":::

Queries contain metadata properties that indicate how the results should be interpreted. For instance, some queries display a simple message at a single location in the code. Others display a series of locations that represent steps along a data-flow or control-flow path, along with a message explaining the significance of the result. Queries that don’t have metadata are not interpreted—their results are output as a table and not displayed in the source code.

Following interpretation, results are output for code review and triaging. In CodeQL for Visual Studio Code, interpreted query results are automatically displayed in the source code. Results generated by the CodeQL CLI can be output into a number of different formats for use with different tools.

4-what-is-ql.md - existing unit

QL is a declarative, object-oriented query language that is optimized to enable efficient analysis of hierarchical data structures, in particular, databases representing software artifacts.

A database is an organized collection of data. The most commonly used database model is a relational model which stores data in tables and SQL (Structured Query Language) is the most commonly used query language for relational databases.

The purpose of a query language is to provide a programming platform where you can ask questions about information stored in a database. A database management system manages the storage and administration of data and provides the querying mechanism. A query typically refers to the relevant database entities and specifies various conditions (called predicates) that must be satisfied by the results. Query evaluation involves checking these predicates and generating the results. Some of the desirable properties of a good query language and its implementation include:

* **Declarative specifications** - a declarative specification describes properties that the result must satisfy, rather than providing the procedure to compute the result. In the context of database query languages, declarative specifications abstract away the details of the underlying database management system and query processing techniques. This greatly simplifies query writing.
* **Expressiveness** - a powerful query language allows you to write complex queries. This makes the language widely applicable.
* **Efficient execution** - queries can be complex and databases can be very large, so it is crucial for a query language implementation to process and execute queries efficiently.

In this unit, you will learn about the basic features of the QL programming language so that you can write your own custom queries or better understand the pre-existing open source queries available.

## The QL syntax

The syntax of QL is similar to SQL, but the semantics of QL are based on Datalog, a declarative logic programming language often used as a query language. This makes QL primarily a logic language, and all operations in QL are logical operations. Furthermore, QL inherits recursive predicates from Datalog, and adds support for aggregates, making even complex queries concise and simple. For example, consider a database containing parent-child relationships for people. If you want to find the number of descendants of a person, typically you would:

1. Find a descendant of the given person, that is, a child or a descendant of a child.
2. Count the number of descendants found using the previous step.

When you write this process in QL, it closely resembles the above structure. Notice that the example used recursion to find all descendants of the given person, and an aggregate to count the number of descendants. Translating these steps into the final query without adding any procedural details is possible due to the declarative nature of the language. The QL code would look something like this:

```ql
Person getADescendant(Person p) {
  result = p.getAChild() or
  result = getADescendant(p.getAChild())
}

int getNumberOfDescendants(Person p) {
  result = count(getADescendant(p))
}
```

## Object orientation

Object orientation is an important feature of QL. The benefits of object orientation are well-known – it increases modularity, enables information hiding, and allows code reuse. QL offers all these benefits without compromising on its logical foundation. This is achieved by defining a simple object model where classes are modeled as predicates and inheritance as implication. The libraries made available for all supported languages make extensive use of classes and inheritance.

## QL and general purpose programming languages

Here are a few prominent conceptual and functional differences between general purpose programming languages and QL:

* QL does not have any imperative features such as assignments to variables or file system operations.
* QL operates on sets of tuples and a query can be viewed as a complex sequence of set operations that defines the result of the query.
* QL’s set-based semantics makes it very natural to process collections of values without having to worry about efficiently storing, indexing and traversing them.

In object-oriented programming languages, instantiating a class involves creating an object by allocating physical memory to hold the state of that instance of the class. In QL, classes are just logical properties describing sets of already existing values.

Exercise - new exercise needed - new unit
12-summary.md - edited existing unit

You are a senior developer responsible for implementing automated code vulnerability scanning at your company. You need to understand how code scanning with CodeQL works and how to configure it, so that you can help your entire organization adopt it.

You did some research on code scanning with CodeQL and found the following:

* Code scanning with CodeQL uses a workflow file that specifies the location of queries, which languages to analyze, and whether they should be built with autobuild, or manual build steps
* GitHub supports integration of third party scanning and alerting tools in the code scanning process
* CodeQL has a CLI that allows you to create and analyze databases offline and then upload the results to GitHub using a SARIF file

Without using GitHub code scanning with CodeQL, it would be very difficult to automate both the scanning of your code, as well as generating pull requests to fix the vulnerable code. In addition, CodeQL provides an extensive, growing library of queries in multiple languages that help you create more secure code with little engineering effort.

Successfully rolling out automated code vulnerability scanning across your organization has made developers more productive and the product at your company more secure.

## References

* [GitHub CodeQL](https://codeql.github.com/)
* [GitHub code scanning](https://docs.github.com/en/code-security/code-scanning)

rmallorybpc commented:
# Code Scanning with GitHub CodeQL

# introduction - for new content - [about integration with code scanning](https://docs.github.com/en/code-security/code-
# scanning/integrating-with-code-scanning/about-integration-with-code-scanning) - new unit

# 5-code-scanning-codeql.md - existing unit

# 9-use-codeql-cli.md - [more content if needed](https://docs.github.com/en/code-security/code-scanning/using-codeql-code-scanning-with-your-existing-ci-system/about-codeql-code-scanning-in-your-ci-system) - existing unit
[Recommended hardware resources for running CodeQL](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/recommended-hardware-resources-for-running-codeql) - new unit
# Exercise - new exercise needed - new unit
# summary.md - edited, revised new unit

rmallorybpc commented:
## Customize Code Scanning with GitHub CodeQL

introduction.md - edited, revised new unit
[Configuring the CodeQL workflow for compiled languages](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-the-codeql-workflow-for-compiled-languages) - new unit
6-customize-your-scanning-workflow-with-codeql.md - existing unit
7-exercise-reference-codeql-query.md - existing unit
8-customize-your-scanning-workflow-with-codeql-2.md - existing unit
11-exercise-configure-language-matrix.md - existing unit
10-custom-build-steps-for-code-scanning.md - existing unit
[Troubleshooting the CodeQL workflow](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/troubleshooting-the-codeql-workflow) - new unit
Exercise(s) needed - new unit
summary.md, revised new unit

camihmerhar commented:
# Feedback on Introduction to Code scanning with GitHub CodeQL

Howdy @rmallorybpc , apologies for the delay on the feedback! First of all great work on all of the CodeQL content! It's extensive!  

I did my first grammatical pass on this on Wednesday, but then I needed a bit more time to do a structural pass. 

Anyway, here are my thoughts, let me know what you think!

**Grammatical Edits:**

**Introduction:**
- [ ] Add an apostrophe in the first sentence in organizations, so the sentence should read as a possessive, "Code scanning using CodeQL provides an extensible method to automate vulnerability scanning across _your organization's_ GitHub repositories."
- [ ] To avoid redundancy in the unit breakdown I would reword the second bullet point to, "Learn about QL, a unique logic programming language"

**Unit 1**
- [ ] Totally up to you, but I would breakdown the introduction of the subunits into a bullet point list to make is easier for learners to read through it. 

- Okay, get ready for me deleting commas right and left. Feel free to disregard. I think I have bad memories of getting points taken off writing assignments because I was taught if you naturally pause in a sentence you should use a comma. But during my junior year of high school my English teacher told me for the most part commas make the most sense when you're:
   1) providing and list of things 
   2) nonrestrictive clauses 
   3) direct addresses (example: John, I think you should stop.)
   4) direct quotes 

This proved to be useful for college, especially as an English major. All of that to say is death to commas. 

- [ ] Remove comma after "...automate security checks..." so the sentence reads like, "CodeQL is the analysis engine used by developers to automate security checks and by security researchers to perform variant analysis."
- [ ] Remove comma after "...engineers use to identify potential vulnerabilities..." so the sentence should read like, "It’s a technique that security engineers use to identify potential vulnerabilities and ensure these threats are properly fixed across multiple codebases."
- [ ] Remove comma after "...during the extraction process" so the sentence should read like, "The schema provides an interface between the initial lexical analysis performed during the extraction process and the actual complex analysis of the CodeQL query evaluator."
- [ ] In the second paragraph of the subunit *Query Suites* I retooled the 2nd sentence to remove the parenthesis, "A suite definition is a sequence of instructions, where each instruction is usually a YAML mapping with a single key." I feel like it flows just a bit more and you don't have to include parentheses.
- [ ] I also added a transition statement to the end of the unit to help with the flow to unit 2, "Next up, you will learn about the three phases of CodeQL analysis."

**Unit 2**

- [ ] Remove comma so the sentence reads like, "To create a database CodeQL first extracts a single relational representation of each source file in the codebase."
- [ ] Remove comma so the sentence reads like, "For compiled languages extraction works by monitoring the normal build process."
- [ ] Rewrote this the sentence for better flow, "For interpreted languages the extractor runs directly on the source code, which resolves dependencies in order to give an accurate representation of the codebase."
- [ ] Remove comma so the sentence reads like, "After extraction all the data required for analysis (relational data, copied source files, and a language-specific database schema, which specifies the mutual relations in the data) is imported into a single directory, known as a CodeQL database."
- [ ] In the last sentence of unit 2, I would change it to say "Next up, you'll learn about QL."

**Unit 3**
- [ ] I would rewrite the last sentence to be a little more colloquial as "Next up, you'll tackle an exercise that'll utilize the knowledge you just learned."

**Summary**
- I feel like you don't need the first sentence in the introductory paragraph of the summary unit. Although you're recalling back to the hypothetical at the beginning, I don't think it fits here because you don't follow through with the example in previous units so it feels out of place. 

**Overall Context and Structural Edits:**

- It might because of my lack of knowledge of CodeQL, but overall I was craving insight into how all of the content related back to CodeQL and its importance to knowing how everything makes organizations more secure. At times, especially in the first unit and 3rd unit, I felt like I was learning terms in a vacuum. One thing that I think might help is providing a sentence within every subunit that ties it back. 
  - For example, I think it would help to have a short sentence after we layout the subunits that way learners can anticipate what they will learn. Like in Unit 1 we tell learners they're going to learn about variant analysis, CodeQL databases, query suites, and query language packs. All great, but why is this important? Although redundant, repetition helps learners retain important concepts. 
  - Another example of helping learners with content is providing a sentence within the subunit that takes it back to why this subject is relevant. I tried to do this in the first unit, in the subunit of **Variant analysis** by taking one of the introduction paragraphs in unit 1 about variant analysis and put it in the the subunit. So now it reads as, 

## Variant analysis

CodeQL is the analysis engine used by developers to automate security checks, and by security researchers to perform variant analysis.

*Variant analysis is the process of using a known security vulnerability as a seed to find similar problems in your code. It’s a technique that security engineers use to identify potential vulnerabilities, and ensure these threats are properly fixed across multiple codebases.*

Querying code using CodeQL is the most efficient way to perform variant analysis. You can use the standard CodeQL queries to identify seed vulnerabilities, or find new vulnerabilities by writing your own custom CodeQL queries. Then, develop or iterate over the query to automatically find logical variants of the same bug that could be missed using traditional manual techniques.

*Overall amazing job, good call breaking this down into multiple modules!*

rmallorybpc commented:
"For example, I think it would help to have a short sentence after we layout the subunits that way learners can anticipate what they will learn. Like in Unit 1 we tell learners they're going to learn about variant analysis, CodeQL databases, query suites, and query language packs. All great, but why is this important? Although redundant, repetition helps learners retain important concepts."

@camihmerhar where else specifically?

rmallorybpc commented:
Pending
customize-code-scanning-codeql - branch
customize-code-scanning-with-github-codeql - folder
https://github.com/githubpartners/microsoft-learn/tree/customize-code-scanning-codeql/github/customize-code-scanning-with-github-codeql

YML files
2-learning-content.yml to 2-configure-codeql.yml
3-learning-content.yml 3-customize-your-scanning-workflow-with-codeql.yml
5-exercise.yml to 5-exercise-reference-codeql-query.yml
4-learning-content.yml to 5a.customize-your-scanning-workflow-with-codeql-2.yml
add new yml file: 5b-exercise-configure-language-matrix.yml
add new yml file: 5c-custom-build-steps-for-code-scanning.yml
add new yml file: 5d-troubleshooting-the-codeql-workflow.yml
6-knowledge-check.yml

MD files
2-learning-content.yml to 2-configure-codeql.md
3-learning-content.yml 3-customize-your-scanning-workflow-with-codeql.md
5-exercise.yml to 5-exercise-reference-codeql-query.md
4-learning-content.yml to 5a.customize-your-scanning-workflow-with-codeql-2.md
add new md file: 5b-exercise-configure-language-matrix.md
add new md file: 5c-custom-build-steps-for-code-scanning.md
add new md file: 5d-troubleshooting-the-codeql-workflow.md
7-summary.md

camihmerhar commented:
Howdy @rmallorybpc! I took another quick pass at the module and here are my thoughts. 

Structurally, I think there are 2 things we can do to make the module more intuitive for learners:

1.  **Ensure that each unit name aligns with the Learning Objectives.**
    - Ultimately the Learning Objectives in the Introduction and Summary should read as:
       - **Learning objectives:**
          - Understanding variant analysis, CodeQL databases, query suites, and query language packs relevancy with Code Scanning
          - How to run a CodeQL analysis
          - What is QL and how to use it with Code Scanning 
2.  **A relevancy statement for each subunit to tie it back how it helps Code Scanning with CodeQL.**
    -  For example for the 2nd Unit, What is CodeQL in the intro paragraph somewhere it should have something like this :
        - In this unit, you will learn about:
           - variant analysis: the process of using a known security vulnerability as a seed to find similar problems in your code
           - CodeQL databases: *insert relevancy statement*
           - query suites: *insert relevancy statement*
           - and query language packs: *insert relevancy statement*

Based of off my experience online content learners/readers only take 7-10 seconds to scan a page, so by these two adjustments to help learners know what to expect it will help them digest information more effectively and easily.

Let me know if you have any questions!

camihmerhar commented:
# Feedback for Analyze Code Using CodeQL

Howdy @rmallorybpc, great work as always! Let me know if you have any questions.

Grammatical and Structural Suggestions:

**Introduction**
- [ ] I feel like we can do away with this sentence, "In this module, you will learn about integrating with code scanning, set up code scanning with GitHub Actions, use the CodeQL command-line interface (CLI), and understand what hardware resources to use with CodeQL." It seems redundant to the learning objectives. 

**Unit 2**
- [ ] Yay table of contents! It looks great, but I'm thinking the short description after the topic looks text heavy. I think we should just have the table of contents with the title of the subunit and then have a strong relevancy/transitional sentence at the beginning of every subunit. Thoughts?
- [ ] Minor edit, but I feel like we can break up the first paragraph in the "Integrate with code scanning" subunit so it reads easier so it'll look like...

"You can perform analysis elsewhere and then upload the code scanning results to GitHub. The alerts for code scanning that you run externally are displayed in the same way as those for code scanning that you run within GitHub. 

When you use a third-party static analysis tool that can produce results as Static Analysis Results Interchange Format (SARIF) 2.1.0 data, you can upload the results to GitHub."

- [ ] Similarly, I feel like we can do the same to the first paragraph subunit "Integrate with Webhooks," so it reads like...

"You can use code scanning webhooks to build or set up integrations that subscribe to code scanning events in your repository, such as GitHub Apps or OAuth Apps. 

For example, you could build an integration that creates an issue on GitHub or sends you a Slack notification when a new code scanning alert is added in your repository."

**Unit 3**

- [ ] In the second to last paragraph of first subunit of the second learning unit (wow that's a mouthful), I would recommend rewording it to have a bit more flow to something along the lines of, "As a result of either of these two occurrences, code scanning will automatically start."

**Unit 4**

- [ ] Add a serial comma in the sentence "This unit will cover using the CodeQL CLI to create databases, analyze databases, and upload the results to GitHub."
- [ ] Delete comma in the sentence "Once you've made the CodeQL CLI available to servers in your CI system and ensured that they can authenticate with GitHub, you're ready to generate data."
- [ ] In the transition statement to the exercise, I would recommend rewording to something along the lines of, "Next up, you'll take on the role of a detective in a hands-on exercise to find the thief in a fictional QL village." I think this will help the readers know they'll be switching gears to something else. 

**Summary**

- [ ] I would recommend removing the first sentence of the opening paragraph, it feels a touch out of place.

rmallorybpc commented:
@camihmerhar 
> **Unit 2**
> * [x]  Yay table of contents! It looks great, but I'm thinking the short description after the topic looks text heavy. I think we should just have the table of contents with the title of the subunit and then have a strong relevancy/transitional sentence at the beginning of every subunit. Thoughts?
> 
I kept in one start sentence, but removed the second sentence. A sentence to launch the module, but shorter.

camihmerhar commented:
# Feedback on Customize code scanning with GitHub CodeQL

## Introduction

- [ ] I would recommend removing the first sentence of the opening paragraph, "Imagine that you continue to be a senior developer at a start-up company specializing in health care software." We don't use it as an example in the units so I feel like we don't need it. 
- [ ] I recommend adding in sub bullets in the learning objectives for the subunits that you include to help learners anticipate what might be coming. An example would be:
- [ ] It looks like we're missing a unit 4, the titles in the md and yaml files go from 3 to 5 , do we just need to rename a file to 4? 

- Configure the language matrix in a CodeQL workflow
- Customize scanning workflow with CodeQL
  - Customize your scanning workflow with CodeQL workflow
  - Custom build steps for code scanning
  - Troubleshooting the CodeQL workflow 
- Implement custom build steps

## Unit 2 Configure CodeQL

- [ ] I would recommend adding an intro statement leading into the unit to help prep learners get into the content, potentially some type of explanation as to why we need to go over QL language before we dive into the title of the unit, Configure CodeQL

## Unit 3 Customize Code Scanning with GitHub 
- [ ] Since the transition statement/ table of contents of the unit has more than 2 topics you'll cover I would bullet out the topics
- [ ] I recommend making the titles of the subunits the exact same as the ones you list in the transition statement/ table of contents that way if learners are skimming to refresh they know in what order and the title they need to look for to get to the content they want 
- [ ] Should the sub sub unit Use CodeQL Query Packs be a sub unit? 

## 5b Exercise 

- [ ] Minor edit, there needs to be a space between the first sentence of the unit and the second sentence. 

## 5d 
- [ ] Totally up to you but I feel like instead of being 5d I feel the content in this unit can be a standalone so it could potentially just be 6 Troubleshooting the CodeQL Workflow

Great work as always @rmallorybpc! Let me know if you have any questions!
